File: src/lib/podExploreTopicPrefetchCache.js
import { canFetchLiveNews } from './podTopicNewsShared';
import { fetchExploreTopicFeed } from './podExploreTopicFeed';

const TTL_MS = 12 * 60 * 1000;
const store = new Map();
const inflight = new Map();

/** Cache key includes region for dual Local/International startup topics. */
export function exploreTopicCacheKey(section, topicId, startupRegion) {
  const r =
    startupRegion === 'local' || startupRegion === 'international' ? startupRegion : 'international';
  return `${section}|${topicId}|${r}`;
}

export function getExploreTopicFeedCache(key) {
  const e = store.get(key);
  if (!e) return null;
  if (Date.now() - e.ts > TTL_MS) {
    store.delete(key);
    return null;
  }
  return e;
}

export function setExploreTopicFeedCache(key, entry) {
  store.set(key, {
    items: entry.items,
    error: entry.error || '',
    ts: Date.now(),
  });
}

export function invalidateExploreTopicFeedCache(key) {
  store.delete(key);
}

/**
 * Warm cache before navigating to /pod/explore/:section/:topicId (default region international).
 */
export function prefetchExploreTopicRaw(section, topicId, startupRegion = 'international') {
  if (!section || !topicId) return Promise.resolve();
  if (!canFetchLiveNews() && section !== 'ai-tech' && section !== 'entrepreneurship' && section !== 'current-affairs')
    return Promise.resolve();
  const key = exploreTopicCacheKey(section, topicId, startupRegion);
  if (getExploreTopicFeedCache(key)) return Promise.resolve();
  if (inflight.has(key)) return inflight.get(key);

  const p = fetchExploreTopicFeed({ section, topicId, startupRegion })
    .then((res) => {
      inflight.delete(key);
      if (res.items?.length) {
        setExploreTopicFeedCache(key, { items: res.items, error: res.error || '' });
      }
      return res;
    })
    .catch(() => {
      inflight.delete(key);
    });

  inflight.set(key, p);
  return p;
}

File: flutter_app/lib/lib/pod_explore_topic_prefetch_cache.dart
import 'pod_explore_topic_feed.dart';
import 'pod_topic_news_shared.dart';

const int _exploreCacheTtlMs = 12 * 60 * 1000;

class ExploreTopicCacheEntry {
  ExploreTopicCacheEntry({
    required this.items,
    required this.ts,
    this.error = '',
  });
  final List<NewsArticle> items;
  final String error;
  final int ts;
}

final _exploreStore = <String, ExploreTopicCacheEntry>{};
final _exploreInflight = <String, Future<void>>{};

String exploreTopicCacheKey(String section, String topicId, String startupRegion) {
  final r = startupRegion == 'local' || startupRegion == 'international'
      ? startupRegion
      : 'international';
  return '$section|$topicId|$r';
}

ExploreTopicCacheEntry? getExploreTopicFeedCacheEntry(String key) {
  final e = _exploreStore[key];
  if (e == null) return null;
  if (DateTime.now().millisecondsSinceEpoch - e.ts > _exploreCacheTtlMs) {
    _exploreStore.remove(key);
    return null;
  }
  return e;
}

List<NewsArticle>? getExploreTopicFeedCache(String key) =>
    getExploreTopicFeedCacheEntry(key)?.items;

void setExploreTopicFeedCache(String key, List<NewsArticle> items, {String error = ''}) {
  _exploreStore[key] = ExploreTopicCacheEntry(
    items: items,
    error: error,
    ts: DateTime.now().millisecondsSinceEpoch,
  );
}

void invalidateExploreTopicFeedCache(String key) {
  _exploreStore.remove(key);
  _exploreInflight.remove(key);
}

Future<void> prefetchExploreTopicRaw(
  String section,
  String topicId, [
  String startupRegion = 'international',
]) async {
  if (section.isEmpty || topicId.isEmpty) return;
  if (!canFetchLiveNews() &&
      section != 'ai-tech' &&
      section != 'entrepreneurship' &&
      section != 'current-affairs') {
    return;
  }
  final key = exploreTopicCacheKey(section, topicId, startupRegion);
  if (getExploreTopicFeedCacheEntry(key) != null) return;
  if (_exploreInflight.containsKey(key)) {
    await _exploreInflight[key];
    return;
  }

  final p = fetchExploreTopicFeed(
    section: section,
    topicId: topicId,
    startupRegion: startupRegion,
  ).then((res) {
    _exploreInflight.remove(key);
    if (res.items.isNotEmpty) {
      setExploreTopicFeedCache(key, res.items, error: res.error);
    }
  }).catchError((_) {
    _exploreInflight.remove(key);
  });

  _exploreInflight[key] = p;
  await p;
}
